hail/python/test/hailtop/batch/test_batch.py

```diff
@@ -88,6 +89,21 @@ def test_single_job(self):
 
             assert self.read(output_file.name) == msg
 
+    @unittest.skipIf(not which("docker"), "docker command is missing")
+    def test_single_job_with_entrypoint(self):
+        with tempfile.NamedTemporaryFile('w') as output_file:
+            msg = 'hello world'
+
+            b = self.batch()
+            j = b.new_job()
+            j.image('hailgenetics/entrypoint-test')
```

Comment:
Could you also delete the `entrypoint-test` image?

Reply:
Sure. Yes, I hadn't done this because nginx is a much heavier image than alpine with bash, so using nginx would be slower. I also didn't find any docker images that we rely on that use entrypoints.

Reply:
Personally I'd build the image during the build step instead of introducing an external dependency.  You have to build the image locally to test locally.  That logic normally goes in a Makefile.  A Makefile for batch tests that builds the entrypoint image seems reasonable.  That image should be built on ubuntu:18.04 like all (most) of our images.  That will always be lighter weight than an external image because it will (almost) always be cached.

Reply:
>You have to build the image locally to test locally

What does this mean? 

Trying to understand how batch/Makefile is used; I don't see where it's called by CI. I do see that I can use a buildImage step; I could just point that to the Dockerfiles I made for this PR (based on ubuntu:18.04, adds entrypoint and /bin/bash), which could live in batch/.

Reply:
Sorry I missed this, thanks for the bump.  The Makefiles are not called by CI, they are parallel constructions, and are primarily in place to bootstrap the cluster.  The Makefiles (roughly) contain the steps needed to install everything up to CI on a fresh cluster so that CI is then able to do a full deploy.  They are also local versions of certain CI build steps.  So if you want to run a build step locally (like a test), I would say that's the natural place.  I don't know if there are existing examples for local tests.  If your step (a test step) depends on another step (like building an image), it would also have to go into the Makefile.

To put this in context, in the future, we might want:
 - A tool that can run some build.yaml steps locally.  That's complicated because some steps need secrets or access to K8s/GCP networking that are not available locally, so the design isn't clear here.  This could the current Makefiles.
 - Fully automate (and test!) spinning up or upgrading a new cluster (probably using Terraform).  This could replace the current Makefiles (or there'd be just a few Terraform make targets).

The Makefiles rot quickly because they are not tested automatically.

 > I can create a buildImage step to build this as a prerequisite to test_hailtop_batch. Is that what you meant?

Yes.  That handles the case of running these tests under CI.  I thought you had concerns about running the tests locally. That's what the Makefile comment was about.

Reply:
Thanks for the explanation. I've dismissed your last review after addressing the shel testing comments. If you prefer that I test entrypoint in this PR as well, I will follow up in the morning, updating creating Dockerfile.entrypoint-test (or use docker build with --build-arg), updating docker/Makefile, batch/Makefile, build.yaml, and adding new tests for local backend and service backend.